### PR TITLE
feat(lsp): add lsp.util.split_lines and lsp.util.split_lines_iter

### PR DESCRIPTION
**Pull request by dmitmel**
How do I generate documentation for `lsp.util.split_lines` and `lsp.util.split_lines_iter`?

### CODE REVIEW

File: runtime/lua/vim/lsp/diagnostic.lua [hunk 1]
```diff
@@ -102,9 +102,10 @@ local function diagnostic_lsp_to_vim(diagnostics, bufnr, client_id)
       end_lnum = _end.line,
       end_col = line_byte_from_position(buf_lines, _end.line, _end.character, offset_encoding),
       severity = severity_lsp_to_vim(diagnostic.severity),
-      message = diagnostic.message,
+      message = table.concat(vim.lsp.util.split_lines(diagnostic.message), '\n'),
```
Comment (dmitmel) [inline runtime/lua/vim/lsp/diagnostic.lua:105]: Should the lines be normalized here, in vim.lsp.diagnostic, or handled by vim.diagnostic according to LSP rules?

Comment (mjlbach) [reply runtime/lua/vim/lsp/diagnostic.lua:105]: This is just the diagnostic message displayed, I don't think we need to split and concatenate here? Is the server actually returning anything other than `\n`?

File: runtime/lua/vim/lsp.lua [hunk 2]
```diff
@@ -115,6 +115,13 @@ local format_line_ending = {
   ["mac"] = '\r',
 }
 
+---@private
+---@param bufnr (number)
+---@returns (string)
+local function buf_get_line_ending(bufnr)
+  return format_line_ending[nvim_buf_get_option(bufnr, 'fileformat')] or '\n'
```
Comment (mjlbach) [inline runtime/lua/vim/lsp.lua:122]: in what scenario do we need the fallback?

File: runtime/lua/vim/lsp/diagnostic.lua [hunk 3]
```diff
@@ -132,7 +133,7 @@ local function diagnostic_vim_to_lsp(diagnostics)
         },
       },
       severity = severity_vim_to_lsp(diagnostic.severity),
-      message = diagnostic.message,
+      message = (diagnostic.user_data and diagnostic.user_data.lsp_original_message) or diagnostic.message,
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/diagnostic.lua:136]: This seems out of scope of this PR?

File: runtime/lua/vim/lsp/util.lua [hunk 4]
```diff
@@ -85,9 +85,95 @@ local function get_border_size(opts)
...
-local function split_lines(value)
-  return split(value, '\n', true)
+--- Splits the given text into lines according to LSP's definition of text
+--- lines. Quoting the specification: "To ensure that both client and server
+--- split the string into the same line representation the protocol specifies
+--- the following end-of-line sequences: '\n', '\r\n' and '\r'."
+---
+---@see |vim.lsp.util.split_lines_iter()|
+---@see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocuments
+function M.split_lines(text, opts)
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/util.lua:95]: Given we don't internally handle `\r` according to the specification, I'm not sure this is correct

**Review by mjlbach — COMMENTED** [review]
I need to think more about these. `get_buf_text` is the only thing I think is 100% necessary or correct.

The issue with `\r` is internally, we can have `\r` *without* a properly tracked newline split.  (so we can have midline `\r` unlike other editors)

File: runtime/lua/vim/lsp/diagnostic.lua (hunk 1 above)
Comment (dmitmel) [reply runtime/lua/vim/lsp/diagnostic.lua:105]: > This is just the diagnostic message displayed, I don't think we need to split and concatenate here?

https://github.com/neovim/neovim/blob/9d6a475cedf93a03214ca9c9bd0673a763994d78/runtime/lua/vim/diagnostic.lua#L1230

> Is the server actually returning anything other than `\n`?

Unclear, the specification doesn't state anything concrete either

File: runtime/lua/vim/lsp.lua (hunk 2 above)
Comment (dmitmel) [reply runtime/lua/vim/lsp.lua:122]: It is simply written defensively, AFAIK it is impossible to set fileformat to an invalid value

File: runtime/lua/vim/lsp/diagnostic.lua (hunk 3 above)
Comment (dmitmel) [reply runtime/lua/vim/lsp/diagnostic.lua:136]: Related to https://github.com/neovim/neovim/pull/16286/files/14923f7ce479cd93cddba258261f1e97240b70f0#diff-2c5241b87ba4cc10dd6d1822a2cfa718b1ea523bb9781da4f92585aff7f6b01f, otherwise vim.lsp.diagnostic.get and friends won't return the same diagnostics that were passed in. This is a hack though.

File: runtime/lua/vim/lsp/util.lua (hunk 4 above)
Comment (dmitmel) [reply runtime/lua/vim/lsp/util.lua:95]: Are you talking about the function or the usages of it?

**Comment by dmitmel** [comment]
@mjlbach Alright, I give up. I'll undo the changes to usages of `lsp.util.split_lines`, but let's at least keep the function itself or move it to a different PR. Don't you need something like that in order to handle in-line `\r`s?

**Comment by dmitmel** [comment]
Do any Language Servers return CRLF newlines in textEdits?

**Comment by dmitmel** [comment]
> but let's at least move the function itself to a different PR. Don't you need something like that in order to handle in-line `\r`s?

Nevermind, you can just `nvim_buf_get_lines` and split each line on \r. The only place where I imagine this function could be useful is for `lsp.util.get_lines` (and a similar one in `lsp.diagnostic` which can be refactored to use `lsp.util.get_lines`), which can't do fileformat detection. Tell me if you need `split_lines` - if not, I am pulling it out.

**Comment by dmitmel** [comment]
@mjlbach What do we do with this PR? (Ref: https://github.com/neovim/neovim/pull/16277#issuecomment-968097978)

**Comment by mjlbach** [comment]
WDYM what do we do? I need to think about some of the thing's bjorn said about our endline handling.

**Comment by dmitmel** [comment]
Well, I mean, should I make it a PR with just a fix for #16231 and move split_lines to a different PR, or keep both in this PR, or something else? I have undone the changes to usage of split_lines because I couldn't find a server which reports back CR or CRLF line endings on hover, signature help, diagnostics or something of that matter, i.e. unrelated to file content tracking.

File: runtime/lua/vim/lsp/util.lua (hunk 4 above)
Comment (mjlbach) [reply runtime/lua/vim/lsp/util.lua:95]: I'm talking about we shouldn't be splitting lines on \r before sending to the server

**Comment by dmitmel** [comment]
Moved the `buf_get_full_text` fix to #16334

**Comment by mfussenegger** [comment]
Closing this because it conflicts with https://github.com/neovim/neovim/issues/25272 and the functions added aren't used by other lsp components
